fl4health/model_bases/autoencoders_base.py

```diff
@@ -0,0 +1,162 @@
+
+    def forward(self, input: torch.Tensor) -> torch.Tensor:
+        mu, logvar = self.encode(input)
+        z = self.sampling(mu, logvar)
+        output = self.decode(z)
+        # The shape of the flattened_output can be later restored by having the training image shape,
+        # or the decoder structure.
+        flattened_output = output.view(output.shape[0], -1)
+        return torch.cat((logvar, mu, flattened_output), dim=1)
+
+
+class ConditionalVAE(AbstractAE):
+    def __init__(
+        self,
+        model_type: AutoEncoderType,
+        num_conditions: int,
+        encoder: nn.Module,
+        decoder: nn.Module,
+    ) -> None:
+        """Initializes the ConditionalVAE class.
+
+        Args:
+            model_type (AutoEncoderType): The model type should be CONDITIONAL_VAE.
+            num_conditions (int): The total number of conditions, which is used to prepare the condition vector.
+            encoder (nn.Module): Encoder module defined by the user.
+            decoder (nn.Module): Decoder module defined by the user.
+        """
+        super().__init__(model_type, encoder, decoder)
+        self.num_conditions = num_conditions
+
+    def cat_input_condition(self, input: torch.Tensor, condition: torch.Tensor) -> torch.Tensor:
+        input_cond = torch.cat((input, condition), dim=-1)
+        return input_cond
+
+    def one_hot(self, condition: torch.Tensor) -> torch.Tensor:
+        # One-hot encoding of the condition
+        condition = F.one_hot(condition.to(torch.int64), self.num_conditions).to(condition.device)
+        return condition
+
+    def maybe_reshape(self, condition_matrix: torch.Tensor) -> torch.Tensor:
+        # Make sure the condition has the right shape.
+        if len(condition_matrix.shape) > 2:
+            # Since the condition is replicated to match the size of the input,
+            #  we need to narrow it back just to one number per sample
+            return condition_matrix[:, 0, 0]
+        else:
+            # Condition was not replicated to match the size of the input.
+            return condition_matrix
+
+    def encode(self, input: torch.Tensor, condition: torch.Tensor) -> Tuple[torch.Tensor, torch.Tensor]:
+        # User can decide how to use the condition in the encoder,
+        # ex: using the condition in the middle layers of encoder.
+        mu, logvar = self.encoder(input, condition)
+        return mu, logvar
+
+    def decode(self, latent_vector: torch.Tensor, condition: torch.Tensor) -> torch.Tensor:
+        latent_vector = self.cat_input_condition(latent_vector, condition)
+        output = self.decoder(latent_vector)
+        return output
+
+    def sampling(self, mu: torch.Tensor, logvar: torch.Tensor) -> torch.Tensor:
+        std = torch.exp(0.5 * logvar)
+        eps = torch.randn_like(std)
+        return eps.mul(std).add_(mu)
+
+    def forward(self, input: torch.Tensor) -> torch.Tensor:
+        x = input[:, :-1]  # Exclude the last item (conditions)
```

Comment:
This assumes that that the conditional input is a single value that needs to be `one_hot` encoded. However, the conditional input can be a lot more general than that. For example, it could be a dense vector of auxiliary inputs. I think we want to try to make this a bit more general. If that means that we don't have a base class that restrictions the structure of the forward and CVAEs have their own class with a `def forward(self, input: torch.Tensor, condition: torch.Tensor) -> torch.Tensor` then that's alright.

Reply:
That is a very good point, I am changing the code to remove the restriction of one_hot encoded condition so that the condition can be any tensor from the client side. Encode and decode forward functions can also get the input and condition separately for more flexibility. However, I think we need to restrict the model(CVAE)'s forward function to just receiving a single input tensor to be able to use this model in all the implemented client classes. One thing we probably can do is to replace input and output types in clients with a dictionary, and then we can pass anything to any model and receive anything from any model (output of the model could then be, reconstruction, mu, and logvar for example). But I assume we don't want to change the underlying clients for this usecase. We can think about a more efficient way to do this after we need this functionality for another model type.